pmlc/target/x86/heatmap.tpl.cc

```diff
@@ -1,9 +1,17 @@
 // Copyright 2018, Intel Corporation
+#include <cassert>
 
 #include "pmlc/target/x86/heatmap.h"
 
 namespace pmlc::target::x86 {
 
+static const unsigned STARTUP_COST = 32;
+static const unsigned SPEC_STENCIL_TUPLES = 1;
+static const unsigned SPEC_STENCIL_INDICES = 3;
+// clang-format off
+// Need the spaces between braces, so the macro replacement logic doesnt trigger.
+static unsigned special_stencils[1][3] = { {64, 16, 3} };
```

Comment:
Why do we need to modify the template? I would think that we could introduce any specializations in normal c++ code outside of the template. Generally, I think we should keep source code templates that turn into real code as small as possible due to syntax highlighting, and like this clang-format thing you had to do.

Also, I'm not up to sync with why we need to add this additional code to the heatmap. Is it to somehow speed up the search by introducing even further constraints?

Reply:
This is a specialization that is for xsmm only. It is needed to allow converting to xsmm one of the first convolution of resnet50. Otherwise we are much slower at run time. (I hope I understood what the question was about re: additional code).